File: apps/api/internal/platform/persistence/verifier.go
package persistence

import (
	"context"
	"errors"
	"fmt"
	"time"
)

const storeCloseTimeout = 2 * time.Second

type persistenceStore interface {
	Name() string
	Verify(context.Context) error
	Close(context.Context) error
}

// VerificationResult identifies one successfully checked persistence store.
type VerificationResult struct {
	Store    string
	Duration time.Duration
}

// Verifier checks an ordered set of stores and owns their cleanup.
type Verifier struct {
	stores []persistenceStore
}

// NewVerifier creates a one-shot verifier for the provided stores.
func NewVerifier(stores ...persistenceStore) *Verifier {
	return &Verifier{stores: stores}
}

// Verify checks each store in order and closes every store in reverse order.
func (verifier *Verifier) Verify(ctx context.Context) (results []VerificationResult, returnedError error) {
	defer func() {
		for index := len(verifier.stores) - 1; index >= 0; index-- {
			closeCtx, cancel := context.WithTimeout(context.Background(), storeCloseTimeout)
			closeErr := verifier.stores[index].Close(closeCtx)
			cancel()
			if closeErr != nil {
				returnedError = errors.Join(
					returnedError,
					fmt.Errorf("close %s persistence connection: %w", verifier.stores[index].Name(), closeErr),
				)
			}
		}
	}()

	results = make([]VerificationResult, 0, len(verifier.stores))
	for _, store := range verifier.stores {
		startedAt := time.Now()
		if err := store.Verify(ctx); err != nil {
			return results, fmt.Errorf("verify %s connectivity: %w", store.Name(), err)
		}
		results = append(results, VerificationResult{Store: store.Name(), Duration: time.Since(startedAt)})
	}
	return results, nil
}

File: apps/ingestion/src/norvii_ingestion/publication/persistence/verifier.py
"""One-shot persistence verification with deterministic resource ownership."""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar, Protocol


class PersistenceVerificationError(RuntimeError):
    """Indicate a service-scoped connectivity or cleanup failure."""


class PersistenceStore(Protocol):
    """Define the narrow behavior consumed by the ingestion verifier."""

    name: ClassVar[str]

    def verify(self) -> None:
        """Authenticate and execute a constant read-only operation."""

    def close(self) -> None:
        """Release the underlying driver resources."""


@dataclass(frozen=True, slots=True)
class VerificationResult:
    """Identify one successfully checked persistence store."""

    store: str


class PersistenceVerifier:
    """Check ordered stores and close every acquired resource in reverse order."""

    def __init__(self, stores: list[PersistenceStore]) -> None:
        self._stores = tuple(stores)

    def verify(self) -> tuple[VerificationResult, ...]:
        """Verify every store once and always release each store."""
        results: list[VerificationResult] = []
        verification_error: PersistenceVerificationError | None = None
        try:
            for store in self._stores:
                try:
                    store.verify()
                except Exception as error:
                    raise PersistenceVerificationError(
                        f"Verify {store.name} connectivity failed."
                    ) from error
                results.append(VerificationResult(store=store.name))
        except PersistenceVerificationError as error:
            verification_error = error
        finally:
            cleanup_error = self._close_stores()

        if verification_error is not None:
            raise verification_error
        if cleanup_error is not None:
            raise cleanup_error
        return tuple(results)

    def _close_stores(self) -> PersistenceVerificationError | None:
        first_error: PersistenceVerificationError | None = None
        for store in reversed(self._stores):
            try:
                store.close()
            except Exception:  # noqa: BLE001 - cleanup must cover every protocol implementation
                if first_error is None:
                    first_error = PersistenceVerificationError(
                        f"Close {store.name} persistence connection failed."
                    )
        return first_error
